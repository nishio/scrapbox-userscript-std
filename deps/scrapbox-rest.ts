export type {
  BadRequestError,
  BaseLine as Line,
  ErrorLike,
  ExportedData,
  GuestUser,
  ImportedData,
  InvalidURLError,
  MemberProject,
  MemberUser,
  NoQueryError,
  NotFoundError,
  NotLoggedInError,
  NotMemberError,
  NotMemberProject,
  NotPrivilegeError,
  Page,
  PageList,
  PageSnapshot,
  ProjectId,
  ProjectResponse,
  ProjectSearchResult,
  SearchedTitle,
  SearchResult,
  SessionError,
  Snapshot,
  TweetInfo,
} from "https://raw.githubusercontent.com/scrapbox-jp/types/0.4.2/rest.ts";
