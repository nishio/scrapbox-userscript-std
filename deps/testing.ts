export * from "https://deno.land/std@0.217.0/testing/asserts.ts";
export * from "https://deno.land/std@0.217.0/testing/snapshot.ts";
