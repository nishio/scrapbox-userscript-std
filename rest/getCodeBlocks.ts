import type { Line } from "../deps/scrapbox-rest.ts";
import { pull } from "../browser/websocket/pull.ts";
import {
  CodeTitle,
  extractFromCodeTitle,
} from "../browser/websocket/_codeBlock.ts";

/** pull()から取れる情報で構成したコードブロックの最低限の情報 */
export interface TinyCodeBlock {
  /** ファイル名 */
  filename: string;

  /** コードブロック内の強調表示に使う言語名 */
  lang: string;

  /** タイトル行 */
  titleLine: Line;

  /** コードブロックの中身（タイトル行を含まない） */
  bodyLines: Line[];

  /** コードブロックの真下の行（無ければ`null`） */
  nextLine: Line | null;

  /** コードブロックが存在するページの情報 */
  pageInfo: { projectName: string; pageTitle: string };
}

/** `getCodeBlocks()`に渡すfilter */
export interface GetCodeBlocksFilter {
  /** ファイル名 */
  filename?: string;
  /** syntax highlightに使用されている言語名 */
  lang?: string;
  /** タイトル行の行ID */
  titleLineId?: string;
}

/** 他のページ（または取得済みの行データ）のコードブロックを全て取得する
 *
 * ファイル単位ではなく、コードブロック単位で返り値を生成する \
 * そのため、同じページ内に同名のコードブロックが複数あったとしても、分けた状態で返す
 *
 * @param target 取得するページの情報（linesを渡せば内部のページ取得処理を省略する）
 * @param filter 取得するコードブロックを絞り込むfilter
 * @return コードブロックの配列
 */
export const getCodeBlocks = async (
  target: { project: string; title: string; lines?: Line[] },
  filter?: GetCodeBlocksFilter,
): Promise<TinyCodeBlock[]> => {
  const lines = await getLines(target);
  const codeBlocks: TinyCodeBlock[] = [];

  let currentCode: CodeTitle & {
    /** 読み取り中の行がコードブロックかどうか */
    isCodeBlock: boolean;
  } = {
    isCodeBlock: false,
    filename: "",
    lang: "",
    indent: 0,
  };
  for (const line of lines) {
    if (currentCode.isCodeBlock) {
      const body = extractFromCodeBody(line.text, currentCode.indent);
      if (body === null) {
        codeBlocks[codeBlocks.length - 1].nextLine = line;
        currentCode.isCodeBlock = false;
        continue;
      }
      codeBlocks[codeBlocks.length - 1].bodyLines.push(line);
    } else {
      const matched = extractFromCodeTitle(line.text);
      if (matched === null) {
        currentCode.isCodeBlock = false;
        continue;
      }
      currentCode = { isCodeBlock: true, ...matched };
      codeBlocks.push({
        filename: currentCode.filename,
        lang: currentCode.lang,
        titleLine: line,
        bodyLines: [],
        nextLine: null,
        pageInfo: {
          projectName: target.project,
          pageTitle: target.title,
        },
      });
    }
  }
  return codeBlocks.filter((codeBlock) => isMatchFilter(codeBlock, filter));
};

/** targetを`Line[]`に変換する */
const getLines = async (
  target: { project: string; title: string; lines?: Line[] },
): Promise<Line[]> => {
  if (target.lines !== undefined) {
    return target.lines;
  } else {
    const head = await pull(target.project, target.title);
    return head.lines;
  }
};

/** コードブロックのフィルターに合致しているか検証する */
const isMatchFilter = (
  codeBlock: TinyCodeBlock,
  filter?: GetCodeBlocksFilter,
): boolean => {
  const equals = (a: unknown, b: unknown) => !a || a === b;
  return (
    equals(filter?.filename, codeBlock.filename) &&
    equals(filter?.lang, codeBlock.lang) &&
    equals(filter?.titleLineId, codeBlock.titleLine.id)
  );
};

/** 行テキストがコードブロックの一部であればそのテキストを、そうでなければnullを返す
 *
 * @param lineText {string} 行テキスト
 * @param titleIndent {number} コードブロックのタイトル行のインデントの深さ
 * @return `lineText`がコードブロックの一部であればそのテキストを、そうでなければ`null`を返す
 */
const extractFromCodeBody = (
  lineText: string,
  titleIndent: number,
): string | null => {
  const matched = lineText.replaceAll("\r", "").match(/^(\s*)(.*)$/);
  if (matched === null || matched.length < 2) {
    return null;
  }
  const indent = matched[1];
  const body = matched[2];
  if (indent.length <= titleIndent) return null;
  return indent.slice(indent.length - titleIndent) + body;
};
