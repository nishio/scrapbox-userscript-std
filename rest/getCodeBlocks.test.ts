/// <reference lib="deno.ns" />

import { Line } from "../deps/scrapbox-rest.ts";
import { assertEquals, assertSnapshot } from "../deps/testing.ts";
import { getCodeBlocks } from "./getCodeBlocks.ts";

// https://scrapbox.io/takker/コードブロック記法
const project = "takker";
const title = "コードブロック記法";
const sample: Line[] = [
  {
    "id": "63b7aeeb5defe7001ddae116",
    "text": "コードブロック記法",
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982645,
    "updated": 1672982645,
  },
  {
    "id": "63b7b0761280f00000c9bc21",
    "text": "ここでは[コードブロック]を表現する[scrapbox記法]を示す",
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982645,
    "updated": 1672982671,
  },
  {
    "id": "63b7b0791280f00000c9bc22",
    "text": "",
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982648,
    "updated": 1672982648,
  },
  {
    "id": "63b7b12b1280f00000c9bc3d",
    "text": "サンプル",
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982826,
    "updated": 1672982828,
  },
  {
    "id": "63b7b12c1280f00000c9bc3e",
    "text": " from [/villagepump/記法サンプル#61dd289e7838e30000dc9cb5]",
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982828,
    "updated": 1672982835,
  },
  {
    "id": "63b7b1261280f00000c9bc26",
    "text": "code:コードブロック.py",
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982821,
    "updated": 1672982822,
  },
  {
    "id": "63b7b1261280f00000c9bc27",
    "text": ' print("Hello World!")',
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982822,
    "updated": 1672982822,
  },
  {
    "id": "63b7b1261280f00000c9bc28",
    "text": "無名コードブロック",
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982822,
    "updated": 1672983021,
  },
  {
    "id": "63b7b1261280f00000c9bc29",
    "text": "code:py",
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982822,
    "updated": 1672982822,
  },
  {
    "id": "63b7b1261280f00000c9bc2a",
    "text": ' print("Hello World!")',
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982822,
    "updated": 1672982822,
  },
  {
    "id": "63b7b1261280f00000c9bc2b",
    "text": "インデントつき",
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982822,
    "updated": 1672982822,
  },
  {
    "id": "63b7b1261280f00000c9bc2c",
    "text": " code:インデント.md",
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982822,
    "updated": 1672982822,
  },
  {
    "id": "63b7b1261280f00000c9bc2d",
    "text": "  - インデント\r",
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982822,
    "updated": 1672982822,
  },
  {
    "id": "63b7b1261280f00000c9bc2e",
    "text": "    - インデント",
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982822,
    "updated": 1672982822,
  },
  {
    "id": "63b7b1261280f00000c9bc2f",
    "text": "言語を強制",
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982822,
    "updated": 1672982822,
  },
  {
    "id": "63b7b1261280f00000c9bc30",
    "text": " code:python(js)",
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982822,
    "updated": 1672982822,
  },
  {
    "id": "63b7b1261280f00000c9bc31",
    "text": '  console.log("I\'m JavaScript");',
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982822,
    "updated": 1672982822,
  },
  {
    "id": "63b7b1261280f00000c9bc32",
    "text": "文芸的プログラミング",
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982822,
    "updated": 1672982825,
  },
  {
    "id": "63b7b1261280f00000c9bc33",
    "text": " 標準ヘッダファイルをインクルード",
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982822,
    "updated": 1672982822,
  },
  {
    "id": "63b7b1261280f00000c9bc34",
    "text": "  code:main.cpp",
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982822,
    "updated": 1672982822,
  },
  {
    "id": "63b7b1261280f00000c9bc35",
    "text": "   #include <iostream>",
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982822,
    "updated": 1672982822,
  },
  {
    "id": "63b7b1261280f00000c9bc36",
    "text": "   ",
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982822,
    "updated": 1672982822,
  },
  {
    "id": "63b7b1261280f00000c9bc37",
    "text": " main函数の定義",
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982822,
    "updated": 1672982822,
  },
  {
    "id": "63b7b1261280f00000c9bc38",
    "text": "  code:main.cpp",
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982822,
    "updated": 1672982822,
  },
  {
    "id": "63b7b1261280f00000c9bc39",
    "text": "   int main() {",
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982822,
    "updated": 1672982822,
  },
  {
    "id": "63b7b1261280f00000c9bc3a",
    "text":
      '     std::cout << "Hello, C++" << "from scrapbox.io" << std::endl;',
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982822,
    "updated": 1672982822,
  },
  {
    "id": "63b7b1261280f00000c9bc3b",
    "text": "   }",
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982822,
    "updated": 1672982822,
  },
  {
    "id": "63b7b1261280f00000c9bc3c",
    "text": "   ",
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982822,
    "updated": 1672982822,
  },
  {
    "id": "63b7b0911280f00000c9bc23",
    "text": "",
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982673,
    "updated": 1672982673,
  },
  {
    "id": "63b7b0911280f00000c9bc24",
    "text": "#2023-01-06 14:24:35 ",
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982673,
    "updated": 1672982674,
  },
  {
    "id": "63b7b0931280f00000c9bc25",
    "text": "",
    "userId": "5ef2bdebb60650001e1280f0",
    "created": 1672982674,
    "updated": 1672982674,
  },
];

Deno.test("getCodeBlocks()", async (t) => {
  await assertSnapshot(
    t,
    await getCodeBlocks({ project, title, lines: sample }),
  );
  await t.step("filename filter", async (st) => {
    const filename = "インデント.md";
    const codeBlocks = await getCodeBlocks({ project, title, lines: sample }, {
      filename,
    });
    const yet = [];
    for (const codeBlock of codeBlocks) {
      yet.push(assertEquals(codeBlock.filename, filename));
    }
    await Promise.all(yet);
    await assertSnapshot(st, codeBlocks);
  });
  await t.step("language name filter", async (st) => {
    const lang = "py";
    const codeBlocks = await getCodeBlocks({ project, title, lines: sample }, {
      lang,
    });
    const yet = [];
    for (const codeBlock of codeBlocks) {
      yet.push(assertEquals(codeBlock.lang, lang));
    }
    await Promise.all(yet);
    await assertSnapshot(st, codeBlocks);
  });
  await t.step("title line ID filter", async (st) => {
    const titleLineId = "63b7b1261280f00000c9bc34";
    const codeBlocks = await getCodeBlocks({ project, title, lines: sample }, {
      titleLineId,
    });
    const yet = [];
    for (const codeBlock of codeBlocks) {
      yet.push(assertEquals(codeBlock.titleLine.id, titleLineId));
    }
    await Promise.all(yet);
    await assertSnapshot(st, codeBlocks);
  });
});
