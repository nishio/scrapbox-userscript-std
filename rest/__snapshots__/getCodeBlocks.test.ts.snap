export const snapshot = {};

snapshot[`getCodeBlocks() 1`] = `
[
  {
    bodyLines: [
      {
        created: 1672982822,
        id: "63b7b1261280f00000c9bc27",
        text: ' print("Hello World!")',
        updated: 1672982822,
        userId: "5ef2bdebb60650001e1280f0",
      },
    ],
    filename: "コードブロック.py",
    lang: "py",
    nextLine: {
      created: 1672982822,
      id: "63b7b1261280f00000c9bc28",
      text: "無名コードブロック",
      updated: 1672983021,
      userId: "5ef2bdebb60650001e1280f0",
    },
    pageInfo: {
      pageTitle: "コードブロック記法",
      projectName: "takker",
    },
    titleLine: {
      created: 1672982821,
      id: "63b7b1261280f00000c9bc26",
      text: "code:コードブロック.py",
      updated: 1672982822,
      userId: "5ef2bdebb60650001e1280f0",
    },
  },
  {
    bodyLines: [
      {
        created: 1672982822,
        id: "63b7b1261280f00000c9bc2a",
        text: ' print("Hello World!")',
        updated: 1672982822,
        userId: "5ef2bdebb60650001e1280f0",
      },
    ],
    filename: "py",
    lang: "py",
    nextLine: {
      created: 1672982822,
      id: "63b7b1261280f00000c9bc2b",
      text: "インデントつき",
      updated: 1672982822,
      userId: "5ef2bdebb60650001e1280f0",
    },
    pageInfo: {
      pageTitle: "コードブロック記法",
      projectName: "takker",
    },
    titleLine: {
      created: 1672982822,
      id: "63b7b1261280f00000c9bc29",
      text: "code:py",
      updated: 1672982822,
      userId: "5ef2bdebb60650001e1280f0",
    },
  },
  {
    bodyLines: [
      {
        created: 1672982822,
        id: "63b7b1261280f00000c9bc2d",
        text: "  - インデント",
        updated: 1672982822,
        userId: "5ef2bdebb60650001e1280f0",
      },
      {
        created: 1672982822,
        id: "63b7b1261280f00000c9bc2e",
        text: "    - インデント",
        updated: 1672982822,
        userId: "5ef2bdebb60650001e1280f0",
      },
    ],
    filename: "インデント.md",
    lang: "md",
    nextLine: {
      created: 1672982822,
      id: "63b7b1261280f00000c9bc2f",
      text: "言語を強制",
      updated: 1672982822,
      userId: "5ef2bdebb60650001e1280f0",
    },
    pageInfo: {
      pageTitle: "コードブロック記法",
      projectName: "takker",
    },
    titleLine: {
      created: 1672982822,
      id: "63b7b1261280f00000c9bc2c",
      text: " code:インデント.md",
      updated: 1672982822,
      userId: "5ef2bdebb60650001e1280f0",
    },
  },
  {
    bodyLines: [
      {
        created: 1672982822,
        id: "63b7b1261280f00000c9bc31",
        text: \`  console.log("I'm JavaScript");\`,
        updated: 1672982822,
        userId: "5ef2bdebb60650001e1280f0",
      },
    ],
    filename: "python",
    lang: "js",
    nextLine: {
      created: 1672982822,
      id: "63b7b1261280f00000c9bc32",
      text: "文芸的プログラミング",
      updated: 1672982825,
      userId: "5ef2bdebb60650001e1280f0",
    },
    pageInfo: {
      pageTitle: "コードブロック記法",
      projectName: "takker",
    },
    titleLine: {
      created: 1672982822,
      id: "63b7b1261280f00000c9bc30",
      text: " code:python(js)",
      updated: 1672982822,
      userId: "5ef2bdebb60650001e1280f0",
    },
  },
  {
    bodyLines: [
      {
        created: 1672982822,
        id: "63b7b1261280f00000c9bc35",
        text: "   #include <iostream>",
        updated: 1672982822,
        userId: "5ef2bdebb60650001e1280f0",
      },
      {
        created: 1672982822,
        id: "63b7b1261280f00000c9bc36",
        text: "   ",
        updated: 1672982822,
        userId: "5ef2bdebb60650001e1280f0",
      },
    ],
    filename: "main.cpp",
    lang: "cpp",
    nextLine: {
      created: 1672982822,
      id: "63b7b1261280f00000c9bc37",
      text: " main函数の定義",
      updated: 1672982822,
      userId: "5ef2bdebb60650001e1280f0",
    },
    pageInfo: {
      pageTitle: "コードブロック記法",
      projectName: "takker",
    },
    titleLine: {
      created: 1672982822,
      id: "63b7b1261280f00000c9bc34",
      text: "  code:main.cpp",
      updated: 1672982822,
      userId: "5ef2bdebb60650001e1280f0",
    },
  },
  {
    bodyLines: [
      {
        created: 1672982822,
        id: "63b7b1261280f00000c9bc39",
        text: "   int main() {",
        updated: 1672982822,
        userId: "5ef2bdebb60650001e1280f0",
      },
      {
        created: 1672982822,
        id: "63b7b1261280f00000c9bc3a",
        text: '     std::cout << "Hello, C++" << "from scrapbox.io" << std::endl;',
        updated: 1672982822,
        userId: "5ef2bdebb60650001e1280f0",
      },
      {
        created: 1672982822,
        id: "63b7b1261280f00000c9bc3b",
        text: "   }",
        updated: 1672982822,
        userId: "5ef2bdebb60650001e1280f0",
      },
      {
        created: 1672982822,
        id: "63b7b1261280f00000c9bc3c",
        text: "   ",
        updated: 1672982822,
        userId: "5ef2bdebb60650001e1280f0",
      },
    ],
    filename: "main.cpp",
    lang: "cpp",
    nextLine: {
      created: 1672982673,
      id: "63b7b0911280f00000c9bc23",
      text: "",
      updated: 1672982673,
      userId: "5ef2bdebb60650001e1280f0",
    },
    pageInfo: {
      pageTitle: "コードブロック記法",
      projectName: "takker",
    },
    titleLine: {
      created: 1672982822,
      id: "63b7b1261280f00000c9bc38",
      text: "  code:main.cpp",
      updated: 1672982822,
      userId: "5ef2bdebb60650001e1280f0",
    },
  },
]
`;

snapshot[`getCodeBlocks() > filename filter 1`] = `
[
  {
    bodyLines: [
      {
        created: 1672982822,
        id: "63b7b1261280f00000c9bc2d",
        text: "  - インデント",
        updated: 1672982822,
        userId: "5ef2bdebb60650001e1280f0",
      },
      {
        created: 1672982822,
        id: "63b7b1261280f00000c9bc2e",
        text: "    - インデント",
        updated: 1672982822,
        userId: "5ef2bdebb60650001e1280f0",
      },
    ],
    filename: "インデント.md",
    lang: "md",
    nextLine: {
      created: 1672982822,
      id: "63b7b1261280f00000c9bc2f",
      text: "言語を強制",
      updated: 1672982822,
      userId: "5ef2bdebb60650001e1280f0",
    },
    pageInfo: {
      pageTitle: "コードブロック記法",
      projectName: "takker",
    },
    titleLine: {
      created: 1672982822,
      id: "63b7b1261280f00000c9bc2c",
      text: " code:インデント.md",
      updated: 1672982822,
      userId: "5ef2bdebb60650001e1280f0",
    },
  },
]
`;

snapshot[`getCodeBlocks() > language name filter 1`] = `
[
  {
    bodyLines: [
      {
        created: 1672982822,
        id: "63b7b1261280f00000c9bc27",
        text: ' print("Hello World!")',
        updated: 1672982822,
        userId: "5ef2bdebb60650001e1280f0",
      },
    ],
    filename: "コードブロック.py",
    lang: "py",
    nextLine: {
      created: 1672982822,
      id: "63b7b1261280f00000c9bc28",
      text: "無名コードブロック",
      updated: 1672983021,
      userId: "5ef2bdebb60650001e1280f0",
    },
    pageInfo: {
      pageTitle: "コードブロック記法",
      projectName: "takker",
    },
    titleLine: {
      created: 1672982821,
      id: "63b7b1261280f00000c9bc26",
      text: "code:コードブロック.py",
      updated: 1672982822,
      userId: "5ef2bdebb60650001e1280f0",
    },
  },
  {
    bodyLines: [
      {
        created: 1672982822,
        id: "63b7b1261280f00000c9bc2a",
        text: ' print("Hello World!")',
        updated: 1672982822,
        userId: "5ef2bdebb60650001e1280f0",
      },
    ],
    filename: "py",
    lang: "py",
    nextLine: {
      created: 1672982822,
      id: "63b7b1261280f00000c9bc2b",
      text: "インデントつき",
      updated: 1672982822,
      userId: "5ef2bdebb60650001e1280f0",
    },
    pageInfo: {
      pageTitle: "コードブロック記法",
      projectName: "takker",
    },
    titleLine: {
      created: 1672982822,
      id: "63b7b1261280f00000c9bc29",
      text: "code:py",
      updated: 1672982822,
      userId: "5ef2bdebb60650001e1280f0",
    },
  },
]
`;

snapshot[`getCodeBlocks() > title line ID filter 1`] = `
[
  {
    bodyLines: [
      {
        created: 1672982822,
        id: "63b7b1261280f00000c9bc35",
        text: "   #include <iostream>",
        updated: 1672982822,
        userId: "5ef2bdebb60650001e1280f0",
      },
      {
        created: 1672982822,
        id: "63b7b1261280f00000c9bc36",
        text: "   ",
        updated: 1672982822,
        userId: "5ef2bdebb60650001e1280f0",
      },
    ],
    filename: "main.cpp",
    lang: "cpp",
    nextLine: {
      created: 1672982822,
      id: "63b7b1261280f00000c9bc37",
      text: " main函数の定義",
      updated: 1672982822,
      userId: "5ef2bdebb60650001e1280f0",
    },
    pageInfo: {
      pageTitle: "コードブロック記法",
      projectName: "takker",
    },
    titleLine: {
      created: 1672982822,
      id: "63b7b1261280f00000c9bc34",
      text: "  code:main.cpp",
      updated: 1672982822,
      userId: "5ef2bdebb60650001e1280f0",
    },
  },
]
`;
